src/server/pkg/metrics/metrics.go

```diff
@@ -12,53 +13,131 @@ import (
+func IncrementUserAction(userID string, action string) {
+	incrementActionChannel <- &incrementUserAction{
+		action: action,
+		user:   userID,
+	}
+}
+
+func (r *Reporter) dbMetrics(metrics *Metrics) {
 	cursor, err := gorethink.Object(
 		"Repos",
-		gorethink.DB(pfsDbName).Table("Repos").Count(),
+		gorethink.DB(r.pfsDbName).Table("Repos").Count(),
 		"Commits",
-		gorethink.DB(pfsDbName).Table("Commits").Count(),
+		gorethink.DB(r.pfsDbName).Table("Commits").Count(),
 		"ArchivedCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Archived": true,
 			},
 		).Count(),
 		"CancelledCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Cancelled": true,
 			},
 		).Count(),
 		"Files",
-		gorethink.DB(pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
+		gorethink.DB(r.pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
 		"Jobs",
-		gorethink.DB(ppsDbName).Table("JobInfos").Count(),
+		gorethink.DB(r.ppsDbName).Table("JobInfos").Count(),
 		"Pipelines",
-		gorethink.DB(ppsDbName).Table("PipelineInfos").Count(),
-	).Run(dbClient)
+		gorethink.DB(r.ppsDbName).Table("PipelineInfos").Count(),
+	).Run(r.dbClient)
 	if err != nil {
 		protolion.Errorf("Error Fetching Metrics:%+v", err)
 	}
 	cursor.One(&metrics)
 }
 
-// ReportMetrics blocks and reports metrics, if modified, to the
-// given kubernetes client every 15 seconds.
-func ReportMetrics(clusterID string, kubeClient *kube.Client, address string, pfsDbName string, ppsDbName string) {
-	dbClient, err := db.DbConnect(address)
-	if err != nil {
-		protolion.Errorf("Error connected to DB when reporting metrics: %v\n", err)
-		return
-	}
+// ReportMetrics blocks and reports metrics every 15 seconds
+func (r *Reporter) ReportMetrics() {
+	reportingTicker := time.NewTicker(time.Second * 15)
 	for {
-		metrics := &Metrics{}
-		dbMetrics(dbClient, pfsDbName, ppsDbName, metrics)
-		externalMetrics(kubeClient, metrics)
-		metrics.ID = clusterID
-		metrics.PodID = uuid.NewWithoutDashes()
-		metrics.Version = version.PrettyPrintVersion(version.Version)
-		reportSegment(metrics)
-		<-time.After(15 * time.Second)
+		select {
+		case incrementAction := <-incrementActionChannel:
+			if userActions[incrementAction.user] == nil {
+				userActions[incrementAction.user] = make(countableActions)
+			}
+			val, ok := userActions[incrementAction.user][incrementAction.action]
+			if ok {
+				userActions[incrementAction.user][incrementAction.action] = val.(uint64) + uint64(1)
+			} else {
+				userActions[incrementAction.user][incrementAction.action] = uint64(0)
+			}
+			break
+		case batchOfActions := <-readBatchOfActionsChannel:
+			batchOfActions.channel <- userActions
+			userActions = make(countableUserActions)
+		case <-reportingTicker.C:
+			r.reportClusterMetrics()
+			r.reportUserMetrics()
+		}
```

Comment:
In general stuff like this is good use case for channels there are a couple things wrong with the way you're using them here.

- Performance. The above `for` loops blocks to send metrics to segment twice via `r.reportClusterMetrics()` and `r.reportUserMetrics()` this could easily take a couple of seconds. During this time calls to `IncrementUserAction` will block and it's totally reasonable that that function could be called as part of every single request. So this code will probably wind up adding seconds to the worst case of every single request we have, a pretty huge performance hit.
- You're over using channels here. You really only need the `incrementAction` and the ticker channel. Everything else seems like over channeling to me.